automated test

Geeker1 opened the pull request:
Currently working on automated testing using selenium, mocha and mockjax.

jbothma commented on .github/workflows/muna-frontend.yml:
```diff
@@ -0,0 +1,52 @@
+
+on:
+  push:
+    branches:
+      - automation-test
```
should we add master here?

Geeker1 replied:
Yes

jbothma replied:
please do so. also, please merge this into test.yml

jbothma commented on .github/workflows/muna-frontend.yml:
```diff
@@ -0,0 +1,52 @@
+        node-version: [12.x]
+
+    steps:
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v1
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install seleniumbase
```
what's python needed for?

Geeker1 replied:
It's needed to install selenium

jbothma commented on .github/workflows/muna-frontend.yml:
```diff
@@ -0,0 +1,52 @@
+    - run: NODE_ENV=test npm run build --if-present
+    - run: |
+        npm run serve-test &
+        sleep 10
```
you have no guarantee that this is enough, but it's also a waste of time. Use a polling check to wait for it to be ready with a timeout, e.g.
```
wget --retry-connrefused --waitretry=1 --read-timeout=10 --timeout=10 --tries=10 -O- http://localhost:.../...
```

Geeker1 replied:
I'll fix this

jbothma commented on src/index.html:
```diff
@@ -27,9 +27,8 @@
   <script src="https://kit.fontawesome.com/95e41092af.js" crossorigin="anonymous"></script>
 <meta name="robots" content="noindex"><meta name="theme-color" content="#0094ff"><link rel="manifest" href="manifest.webmanifest"><script>dataLayer = [];</script>
     <script>
-      const pushpadProjectId = `${process.env.PUSHPAD_PROJECT_ID}`;
```
why is this file changing?

jbothma commented on src/js/api.js:
```diff
@@ -2,10 +2,19 @@ const defaultBaseUrl = process.env.DEFAULT_API_URL
   return sessionStorage.getItem("apiBaseUrl") || defaultBaseUrl;
 }
 
+// import { createServer } from "miragejs"
+
+// createServer({
+//   routes() {
+//     this.post("https://muni-portal-backend.openup.org.za/accounts/login/", () => {detail:"Login or password invalid."})
+//   },
+// })
+
+
```
please make sure all commented code is removed when you're ready for a proper review

jbothma commented on src/js/components/account/login.js:
```diff
@@ -70,6 +70,7 @@ export class Login {
    */
   login(endPoint, $form, $success, $fail) {
     const api = new API();
+    // const api = new API();
```
please clean up these things when ready for proper review

jbothma commented on src/js/components/account/login.js:
```diff
@@ -91,6 +92,7 @@ export class Login {
         let errorMessage = "Error while communicating with the server";
         if (jqXHR.status === 400) {
           errorMessage = jqXHR.responseJSON.detail;
+          console.log(errorMessage, jqXHR.responseJSON)
```
please clean up these things when ready for proper review

jbothma commented on src/js/index.js:
```diff
@@ -48,11 +51,19 @@ if (
   );
 }
 
+if(NODE_ENV === 'test'){
+  const {loadAPI} = require('./mocks/mockapi');
+  const mockjax = require('jquery-mockjax');
+
+  mockjax($, window);
+  loadAPI();
+}
+
```
we shouldn't have test code in production code - what's the purpose of this? We should configure the API endpoint used by the app using the environment variable DEFAULT_API_URL - see the README

Geeker1 replied:
In test environment, I added this to mock the api requests so I can test with selenium, DEFAULT_API_URL is used but it's called inside the `loadAPI` function.

jbothma replied:
What I'm saying is we don't want code like `if(NODE_ENV === 'test'){` in a production module (index.js)

Now I can see that to use mockjax, you need to execute javascript in the client context.

I think for end-to-end tests like there, a mock server would be more appropriate than intercepting ajax requests in jquery.

So you'd spin up a mock server, and build the client with `DEFAULT_API_URL=http://localhost:mockport yarn build` or alternatively open the page with ?promptapi in the URL like http://localhost:8080/services/?promptapi and have selenium fill that in if it can.

jbothma commented on src/js/mocks/constant.js:
```diff
@@ -0,0 +1,4 @@
+export const VALID_USER = 'user';
```
can this go under src/js/test/mocks instead?

Geeker1 replied:
ok

jbothma commented on test/app.test.js:
```diff
@@ -0,0 +1,254 @@
+
+	before(async () => {
+		let options = new Options()
+		driver = new webdriver.Builder().forBrowser('firefox')
```
rather use chrome - firefox isn't well-supported right now

Geeker1 replied:
Ok

jbothma commented on test/service-worker.test.js:
```diff
@@ -0,0 +1,34 @@
+*/
+
+
+describe('Service Worker Tests', function() {
```
what did you have in mind for service worker tests? I think you can delete this file for now - let's just finish the user account tests for now

jbothma commented on package-lock.json:
```diff
@@ -0,0 +1,14217 @@
+{
```
Please use yarn rather than npm to add packages - then there shouldn't be any package-long.json and yarn.lock will update

jbothma commented:
Please pull in my changes in the branch shown in https://github.com/OpenUpSA/muni-portal-frontend/pull/118

Also please take a look at what I changed

- I replaced the npm commands to use yarn - I think it's best not to mix npm and yarn if possible
- I removed the python stuff - chrome (and perhaps selenium?) is installed in github actions by default
- I merged master

jbothma commented on .github/workflows/muna-frontend.yml:
```diff
@@ -0,0 +1,54 @@
+    - run: npm ci
+    - run: npm run build:test --if-present
+    - run: |
+        npm run serve-test &
+        npm run mock-server &
+        curl http://localhost:3004 -o /dev/null --connect-timeout 20 --retry 10 --retry-connrefused
+        curl http://localhost:3000 -o /dev/null --connect-timeout 20 --retry 10 --retry-connrefused
+        npm test
```
can you use a node http server controlled from the test suite rather than the command line? then we can locally run tests and have the mock code update each time we run the test without having to look for the process id to kill it and restart it.

How do people normally run mock backend APIs for end-to-end tests?

jbothma replied:
I don't quickly see a javascript API for http-server but there must be something like that

jbothma replied:
I'm happy to merge before we fix this, but if it's quick to solve this now, it'd be great

Geeker1 replied:
> can you use a node http server controlled from the test suite rather than the command line? then we can locally run tests and have the mock code update each time we run the test without having to look for the process id to kill it and restart it.
> 
> How do people normally run mock backend APIs for end-to-end tests?

I don't see a javascript api for http-server

Geeker1 replied:
> How do people normally run mock backend APIs for end-to-end tests?

The options I know of are

1. MirageJS
2. Mockjax (For ajax requests)
3. Jest
4. Json-Server (What I currently use)

Geeker1 replied:
Hi, @jbothma, I have successfully moved mock-server implementation to mocha test code, it works locally, pushed to github actions, build runs successfully.

Geeker1 commented:
> Please pull in my changes in the branch shown in #118
> 
> Also please take a look at what I changed
> 
> * I replaced the npm commands to use yarn - I think it's best not to mix npm and yarn if possible
> * I removed the python stuff - chrome (and perhaps selenium?) is installed in github actions by default
> * I merged master

I've pulled in your changes and pushed to branch.

jbothma commented on src/js/mocks/server.js:
```diff
@@ -0,0 +1,105 @@
+const jsonServer = require('json-server')
+const fs = require('fs');
+const server = jsonServer.create()
+const router = jsonServer.router('db.json')
```
this creates an empty file which just lies around. Can you just pass it an empty object here instead?

jbothma commented on src/js/mocks/server.js:
```diff
@@ -0,0 +1,105 @@
+server.use(jsonServer.bodyParser)
+
+server.post('/api/accounts/login', (req, res) => {
+	const response = {detail: "Login or password invalid."}
```
better to call this errorResponse so it's clear it's not the only response from this mock

jbothma commented on src/js/mocks/server.js:
```diff
@@ -0,0 +1,105 @@
+	)
+})
+
+server.get('/api/wagtail/v2/pages/', (req, res)=>{
+	try{
+		const content = require('./services.json');
+		res.status(200).json(content)
+	} catch{
+		res.status(500).json({error: 'Error reading content.'})
+	}
+})
```
could you apply the linter to your code? for example here is tabs instead of spaces, and there should be a space between `try` and `{`

jbothma commented on src/js/mocks/services.json:
```diff
@@ -0,0 +1,799 @@
+{
+    "meta": {
+        "total_count": 21
+    },
+    "items": [
+        {
+            "id": 9,
+            "meta": {
+                "type": "core.ServicePage",
+                "detail_url": "https://muni-portal-backend.openup.org.za/api/wagtail/v2/pages/9/",
+                "html_url": "https://app.capeagulhas.gov.za/services/town-planning-and-building-control/",
+                "slug": "town-planning-and-building-control",
+                "show_in_menus": false,
+                "seo_title": "",
+                "search_description": "",
```
please cut this down to the minimal data needed to test the feature

jbothma commented on test/app.test.js:
```diff
@@ -0,0 +1,279 @@
+
+const assert = require('assert');
+
+const { server } = require('../src/js/mocks/server');
```
let's call this mockServer since there's also the app server from `serve`

jbothma commented on test/app.test.js:
```diff
@@ -0,0 +1,279 @@
+
+		before(async function(){
+			await driver.get(LOGIN_URL);
+			await driver.sleep(500);
```
never use sleep - always let selenium wait for some element to be present. e.g. on the login page, wait for one of the form elements to be present. the webdriver library might have that built in

jbothma replied:
here's how https://stackoverflow.com/a/43263908/1305080

jbothma commented on test/app.test.js:
```diff
@@ -0,0 +1,279 @@
+			usernameBox.sendKeys(nm);
+			passwordBox.sendKeys(p1);
+
+			const button = driver.findElement(By.className('button form-submit w-button'));
```
you should scope selection of buttons to the modal because there might be more buttons like this hidden in the component library. It might be easier then to select using CSS selector rather than class name.

jbothma commented on test/app.test.js:
```diff
@@ -0,0 +1,279 @@
+			await sendData(VALID_USER, VALID_PASSWORD);
+
+			const url = await driver.getCurrentUrl();
+			const condition = url.endsWith('/services/');
```
could you rather check that the menu contains "logout"? That's a more reliable test than that you were redirected to /services/

jbothma commented on test/app.test.js:
```diff
@@ -0,0 +1,279 @@
+		});
+
+		it('verify logout worked correctly', async function() {
+			const button = driver.findElement(By.className('icon nav-menu__icon'));
```
you need to use something specific to the logout button - can you select by text content?

jbothma commented on test/app.test.js:
```diff
@@ -0,0 +1,279 @@
+			confirmPassword.clear();
+		});
+
+		async function sendData(ol, pa, cf){
```
please use full variable names

jbothma commented on test/app.test.js:
```diff
@@ -0,0 +1,279 @@
+			password.sendKeys(pa);
+			confirmPassword.sendKeys(cf);
+
+			button.click();
```
depending on this variable being set outside this function leads to very broken code - data should be passed as arguments or fetched from the environment inside functions
